pallets/contracts/src/lib.rs

```diff
@@ -13,441 +13,681 @@
+        perms: Permissions,
+    ) -> DispatchResultWithPostInfo {
+        Self::general_instantiate(
+            origin,
+            endowment,
+            // Compute the base weight of roughly `base_instantiate`.
+            <T as Config>::WeightInfo::instantiate_with_hash_bytes(&salt),
+            gas_limit,
+            code_hash,
+            Code::Existing(code_hash),
+            inst_data,
+            salt,
+            perms,
+        )
+    }
 
-        /// Transfer ownership of the template, Can only be called by the owner of template.
-        /// `new_owner` should posses the valid CDD claim.
-        ///
-        /// # Arguments
-        /// * origin Owner of the provided code_hash.
-        /// * code_hash Unique identifer of the template.
-        /// * new_owner Identity that will be the new owner of the provided code_hash.
-        #[weight = <T as Trait>::WeightInfo::transfer_template_ownership()]
-        pub fn transfer_template_ownership(origin, code_hash: CodeHash<T>, new_owner: IdentityId) -> DispatchResult {
-            // Ensure whether the extrinsic is signed & validate the `code_hash`.
-            let (did, _) = Self::ensure_signed_and_template_exists(origin, code_hash)?;
-
-            // Ensuring the `new_owner` is CDD or not.
-            ensure!(Identity::<T>::has_valid_cdd(new_owner), Error::<T>::NewOwnerIsNotCDD);
-
-            // Change the `owner` variable value to the given did.
-            <TemplateInfo<T>>::mutate(&code_hash, |template_details| template_details.owner = new_owner);
-
-            // Emit event.
-            Self::deposit_event(Event::<T>::TemplateOwnershipTransferred(did, code_hash, new_owner));
-            Ok(())
-        }
+    /// General logic for contract instantiation both when the code or code hash is given.
+    ///
+    /// The interesting parameters here are:
+    /// - `base_weight` assigns a base non-variable weight to the full extrinsic.
+    /// - `code_hash` specifies the hash to use to derive the contract's key.
+    /// - `code` specifies the code for the contract, either as a code blob or an existing hash.
+    ///   The hash of `code` in either case is assumed to correspond to `code_hash`.
+    fn general_instantiate(
+        origin: T::Origin,
+        endowment: Balance,
+        base_weight: Weight,
+        gas_limit: Weight,
+        code_hash: CodeHash<T>,
+        code: Code<CodeHash<T>>,
+        inst_data: Vec<u8>,
+        salt: Vec<u8>,
+        perms: Permissions,
+    ) -> DispatchResultWithPostInfo {
+        // Ensure we have perms + we'll need sender & DID.
+        let PermissionedCallOriginData {
+            primary_did: did,
+            sender,
+            ..
+        } = Identity::<T>::ensure_origin_call_permissions(origin)?;
+
+        with_transaction(|| {
+            // Roll back `prepare_instantiate` if contract was not instantiated.
+            Self::prepare_instantiate(did, &sender, &code_hash, &salt, perms)?;
```

Comment:
This would allow any secondary key that is allowed to instantiate a contract to bypass their own limited permissions.